Initial automation

orelmisan opened the pull request:
Add automation for:
- Building the checkup's binary and image
- Unit testing the checkup
- Linting the checkup

To run all targets:
```bash
make
```

To build the checkup image:
```bash
make build
```

To unit test the code:
```bash
make unit-test
```

To lint the code:
```bash
make lint
```

~~Depends on PR #1.~~

orelmisan committed:
Makefile: Add build target

Build the project using the `make` utility.

The container engine binary could be change using the following command:
```bash
CONTAINER_ENGINE=docker make build
```

The image tag could be changed using the following command:
```bash
CHECKUP_IMAGE_TAG=main make build
```

Also discard the multi-stage build Dockerfile.

Signed-off-by: Orel Misan <omisan@redhat.com>

orelmisan committed:
Makefile: Add unit-test target

```bash
make unit-test
```

Signed-off-by: Orel Misan <omisan@redhat.com>

orelmisan committed:
Makefile: Add lint target

```bash
make lint
```

Signed-off-by: Orel Misan <omisan@redhat.com>

orelmisan committed:
Makefile: Add all target

This target runs the lint, unit-test and build targets.

```bash
make
```

Signed-off-by: Orel Misan <omisan@redhat.com>

ormergi commented on Makefile:
```diff
@@ -3,6 +3,15 @@ CONTAINER_ENGINE ?= podman
 .PHONY: build
+
+unit-test:
+	$(CONTAINER_ENGINE) run --rm -v $(PWD):$(PROJECT_WORKING_DIR):Z --workdir $(PROJECT_WORKING_DIR) $(GO_IMAGE_NAME):$(GO_IMAGE_TAG) go test ./...
```
Passing '-v' to Go test framework may come in handy on debugging.
If you don't want to pass it by default, I think its possible to allow passing any flag to the test framework like so:
```
$(CONTAINER_ENGINE) run --rm -v $(PWD):$(PROJECT_WORKING_DIR):Z --workdir $(PROJECT_WORKING_DIR) $(GO_IMAGE_NAME):$(GO_IMAGE_TAG) go test "$@" ./...

make unit-test -v
```

orelmisan replied:
Its a good idea to add the `-v` flag.
I'll add it.

orelmisan replied:
Done.

ormergi commented on Makefile:
```diff
@@ -6,6 +6,9 @@ CHECKUP_IMAGE_TAG ?= devel
 GO_IMAGE_NAME := docker.io/library/golang
 GO_IMAGE_TAG := 1.19.4-alpine3.17
 
+LINTER_IMAGE_NAME := docker.io/golangci/golangci-lint
+LINTER_IMAGE_TAG := v1.50.1
+
```
Shouldn't it be part of the builder image?

orelmisan replied:
There is no `builder` image per se.
The official go image is being used to unit-test and build the checkup's binary.
The official `golangci-lint` image is being used to lint the image.

This approach saves us the need to maintain a `builder` image with all the tooling.

ormergi commented on Makefile:
```diff
@@ -3,6 +3,15 @@ CONTAINER_ENGINE ?= podman
 CHECKUP_IMAGE_NAME := quay.io/kiagnose/kubevirt-rt-checkup
 CHECKUP_IMAGE_TAG ?= devel
 
+GO_IMAGE_NAME := docker.io/library/golang
+GO_IMAGE_TAG := 1.19.4-alpine3.17
```
Shouldn't we use the builder image?

orelmisan replied:
This is the same image as the `builder` image.
I'm considering not to use a multi-stage build, in order to have this image only in one place and instead of two.

orelmisan replied:
I've removed the multi-stage build from the Dockerfile.

orelmisan commented:
[Change](https://github.com/kiagnose/kubevirt-rt-checkup/compare/3933f643614a9d27672a89994beb437c6058044f..99970700003194a335b297e7b2e357e79053bc2b): Added `-v` flag to unit tests.

[Change](https://github.com/kiagnose/kubevirt-rt-checkup/compare/99970700003194a335b297e7b2e357e79053bc2b..6f84d09b07e79f087b9aab728cc664ca9468e2e3): Removed redundant newline.

orelmisan commented:
Rebase

orelmisan commented:
[Change](https://github.com/kiagnose/kubevirt-rt-checkup/compare/72506f29ceda523dbdac23902b7ac4692b6703c5..37146c4e53d53e2a5ebd885e4548fac7352d72ba): Removed the multi-stage build from the Dockerfile in order to keep the go image in one place (Makefile).

ormergi reviewed (dismissed):
LGTM

RamLavi reviewed (changes requested):
can you add push?

orelmisan commented:
> can you add push?

I was thinking to add it with the appropriate github action in a follow-up PR.

Edit: Please see PR #5.

orelmisan commented:
Rebase

EdDev reviewed (approved):
I'm just repeating my other PR comments for the record (so no one will think I was in favor of this :) ).

> It looks nice, although I am not in favor of using makefiles and container images when not a must.
> Taking a tool like makefile and using it for simple scripting does not compute right for me.
> The container image to do all targets is also limiting, as the CI already has tools to test on many go versions. In the current state, you will need to update frequently the config to choose the latest version.
> 
> But this is for you to choose at the end.

Looks good, thanks!
